### PR TITLE
Remove Box from PyObject

### PR DESCRIPTION
**Pull request by OddCoincidence**

### CODE REVIEW

**Comment by coolreader18** [comment]
Ah, it's the same thing again with WASM `held_rcs`.

**Comment by OddCoincidence** [comment]
I always forget about WASM :sweat_smile:

File: wasm/lib/src/vm_class.rs [hunk 1]
```diff
@@ -24,7 +24,7 @@ pub(crate) struct StoredVirtualMachine {
     pub scope: RefCell<Scope>,
     /// you can put a Rc in here, keep it as a Weak, and it'll be held only for
     /// as long as the StoredVM is alive
-    held_rcs: RefCell<Vec<Rc<dyn HeldRcInner>>>,
+    held_objects: RefCell<Vec<PyObjectRef>>,
```
Comment (coolreader18) [inline wasm/lib/src/vm_class.rs]: Yeah, this is probably a better system than what I had. You can just remove `HeldRcInner` now.

Comment (coolreader18) [inline wasm/lib/src/vm_class.rs]: Yeah, this is a better system than what I had. You can just remove `HeldRcInner` now.

File: vm/src/pyobject.rs [hunk 2]
```diff
@@ -788,9 +791,30 @@ pub trait IdProtocol {
     }
 }
 
-impl IdProtocol for PyObjectRef {
+#[derive(Debug)]
+enum Never {}
```
Comment (coolreader18) [inline vm/src/pyobject.rs]: What's the use of this?

Comment (OddCoincidence) [reply vm/src/pyobject.rs]: We need a concrete payload type to cast a fat pointer (`*const PyObject<dyn PyObjectPayload>`) to a thin pointer (`*const PyObject<SomeConcreteType>`) so we can use its address for the object id. The ideal type would be the never type `!`, but that's not stable, so creating an enum with no variants is the standard workaround.

**Pull request merged**